src/Razor/Razor.Runtime/src/RazorRazorPageBase.cs

```diff
@@ -0,0 +1,505 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Html;
+using Microsoft.AspNetCore.Razor.Runtime.TagHelpers;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+
+namespace Microsoft.AspNetCore.Razor
+{
+    /// <summary>
+    /// Represents properties and methods that are needed in order to render a view that uses Razor syntax.
+    /// </summary>
+    public abstract class RazorRazorPageBase<TOutputContext> : IRazorRazorPage<TOutputContext> where TOutputContext : IOutputContext
+    {
+        private readonly Stack<TextWriter> _textWriterStack = new Stack<TextWriter>();
+        private StringWriter _valueBuffer;
+        private AttributeInfo _attributeInfo;
+        private TagHelperAttributeInfo _tagHelperAttributeInfo;
+
+        public virtual TOutputContext ViewContext { get; set; }
+
+        public string Layout { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="TextWriter"/> that the page is writing output to.
+        /// </summary>
+        /// <summary>
+        /// Gets the <see cref="TextWriter"/> that the page is writing output to.
+        /// </summary>
+        public virtual TextWriter Output
+        {
+            get
+            {
+                var viewContext = ViewContext;
+                if (viewContext == null)
+                {
+                    throw new InvalidOperationException("View context must be set"); //TODO Resources.FormatViewContextMustBeSet(nameof(ViewContext), nameof(Output)));
+                }
+
+                return viewContext.Writer;
+            }
+        }
+
+        /// <inheritdoc />
+        public string Path { get; set; }
+
+        /// <inheritdoc />
+        public IDictionary<string, RenderAsyncDelegate> SectionWriters { get; } =
+            new Dictionary<string, RenderAsyncDelegate>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Gets the dynamic view data dictionary.
+        /// </summary>
+        public dynamic ViewBag => ViewContext?.ViewBag;
```

Comment:
I'd really like to avoid having `ViewBag` here. There's an impactful runtime/startup cost to having dynamic types appear.